Cubical/Algebra/IntegerMatrix/Diagonalization.agda

```diff
@@ -0,0 +1,283 @@
+
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.Function
+open import Cubical.Foundations.HLevels
+
+open import Cubical.Data.Nat
+  hiding (_·_) renaming (_+_ to _+ℕ_ ; +-assoc to +Assocℕ)
+open import Cubical.Data.Nat.Order
+open import Cubical.Data.Nat.Divisibility
+  using    (m∣n→m≤n)
+  renaming (∣-trans to ∣ℕ-trans ; ∣-refl to ∣-reflℕ)
+open import Cubical.Data.Int
+  hiding (_+_ ; _·_ ; _-_ ; -_ ; addEq)
+open import Cubical.Data.Int.Divisibility
+open import Cubical.Data.FinData
+
+open import Cubical.Data.Empty as Empty
+open import Cubical.Data.Unit  as Unit
+open import Cubical.Data.Sum
+open import Cubical.Data.Sigma
+open import Cubical.Data.List
+
+open import Cubical.Relation.Nullary
+
+open import Cubical.Algebra.Matrix
+open import Cubical.Algebra.Matrix.CommRingCoefficient
+open import Cubical.Algebra.Matrix.RowTransformation
+open import Cubical.Algebra.Matrix.Elementaries
+open import Cubical.Algebra.IntegerMatrix.Base
+open import Cubical.Algebra.IntegerMatrix.Elementaries
+
+open import Cubical.Algebra.CommRing
+open import Cubical.Algebra.CommRing.Instances.Int
+  renaming (ℤ to ℤRing)
+
+open import Cubical.Induction.WellFounded
+
+private
+  variable
+    m n k : ℕ
+
+open CommRingStr      (ℤRing .snd)
+
+open Coefficient ℤRing
+open Sim
+
+open ElemTransformation ℤRing
+open ElemTransformationℤ
+open SwapPivot
+open RowsImproved
+open ColsImproved
+
+-- Sequence of non-zero integers
+
+isNonZero : List ℤ → Type
+isNonZero [] = Unit
+isNonZero (x ∷ xs) = (¬ x ≡ 0) × isNonZero xs
+
+isPropIsNonZero : (xs : List ℤ) → isProp (isNonZero xs)
+isPropIsNonZero [] = isPropUnit
+isPropIsNonZero (x ∷ xs) = isProp× (isPropΠ (λ _ → isProp⊥)) (isPropIsNonZero xs)
+
+NonZeroList : Type
+NonZeroList = Σ[ xs ∈ List ℤ ] isNonZero xs
+
+cons : (n : ℤ)(xs : NonZeroList) → ¬ n ≡ 0 → NonZeroList
+cons n (xs , _) _ .fst = n ∷ xs
+cons n ([] , _) p .snd = p , tt
+cons n (x ∷ xs , q) p .snd = p , q
+
+-- Smith normal matrix
+
+_+length_ : NonZeroList → ℕ → ℕ
+xs +length n = length (xs .fst) +ℕ n
+
+diagMat : (xs : List ℤ)(m n : ℕ) → Mat (length xs +ℕ m) (length xs +ℕ n)
+diagMat [] _ _ = 𝟘
+diagMat (x ∷ xs) _ _ = x ⊕ diagMat xs _ _
+
+diagMat⊕ :
+    (a : ℤ)(xs : NonZeroList){m n : ℕ}
+  → (p : ¬ a ≡ 0)
+  → a ⊕ diagMat (xs .fst) m n ≡ diagMat (cons a xs p .fst) m n
+diagMat⊕ _ _ _ = refl
+
+-- Diagonal matrix with non-zero diagonal elements
+-- Notice that we allow non-square matrices.
+
+record isDiagonal (M : Mat m n) : Type where
+  field
+    divs : NonZeroList
+    rowNull : ℕ
+    colNull : ℕ
+
+    rowEq : divs +length rowNull ≡ m
+    colEq : divs +length colNull ≡ n
+    matEq : PathP (λ t → Mat (rowEq t) (colEq t)) (diagMat (divs .fst) rowNull colNull) M
+
+open isDiagonal
+
+row col : {M : Mat m n} → isDiagonal M → ℕ
+row isNorm = isNorm .divs +length isNorm .rowNull
+col isNorm = isNorm .divs +length isNorm .colNull
+
+isDiagonal𝟘 : isDiagonal (𝟘 {m = m} {n = n})
```

Comment:
Wouldn't this and the below two definitions be nicer with copattern matching instead of explicit records?

Reply:
done